README.md
# GoSpeccy - An evolving ZX Spectrum 48k Emulator

GoSpeccy is a free ZX Spectrum (Speccy for friends) emulator
written in [Go](http://golang.org).

# Quick start

Installing and starting GoSpeccy with Go1 is simple:

    go get -v github.com/remogatto/gospeccy/...
    gospeccy
    gospeccy -wos="interlace demo"

# Description

GoSpeccy is based on a [concurrent](http://github.com/remogatto/gospeccy/wiki/Architecture) architecture.
We think the concurrency is a strong peculiarity of GoSpeccy as it opens new
interesting scenarios when developing and using the emulator.

Go has interesting features that help in emulators development:
 
* it has "low-level" similarities with C while allowing more productivity
  than C in certain cases

* it is strongly typed and type safe so you are aware about certain errors at
  compile-time

* it is garbage collected so there is small chance of memory leaks

* it has an uint16 built-in type that helps dealing with 8/16 bit
  emulation

* it has goroutines to enable concurrency in the emulator design
  and implementation

The Zilog Z80 CPU emulation is the core of GoSpeccy. The CPU emulation
code is generated using a modified version of the <tt>z80.pl</tt> script
shipped with [FUSE](http://fuse-emulator.sourceforge.net/) (one of the best ZX
Spectrum emulators around). The script has been hacked to generate Go
code rather than C code.

Another source of inspiration was
[JSSpeccy](http://matt.west.co.tt/spectrum/jsspeccy/), a neat
Javascript Speccy emulator.

The Z80 emulation is tested against the excellent test-suite shipped
with FUSE.

If you like this software, please watch it on
[github](http://github.com/remogatto/gospeccy)! Seeing a growing
number of watchers is an excellent motivation for the GoSpeccy team to
keep up this work :) Bug reports and testing are also appreciated! And
don't forget to fork and send patches, of course ;)

# Features

* Complete (almost) Zilog Z80 emulation
* Concurrent [architecture](http://github.com/remogatto/gospeccy/wiki/Architecture)
* Beeper support
* Initial support for Kempston joysticks
* An interactive on-screen console interface based on [clingon](http://github.com/remogatto/clingon)
* Snapshot support: SNA, Z80 formats (48k versions)
* Tape support (TAP format, read-only)
* Accelerated tape loading
* ZIP files support
* SDL backend
* 2x scaler and fullscreen

# Using GoSpeccy

To make the screen bigger try the "-2x" command line option,
or type "scale(2)" in the interactive console.

To try the classic Hello World try to press the following keys:

    p
    CTRL+p
    hello world
    CTRL+p
    RETURN

For a nice picture of the speccy keyboard layout visit this
[page](http://www.guybrush.demon.co.uk/spectrum/docs/Basic.htm).

To load a program run:

    gospeccy file.tap

To enable tape loading acceleration use the <tt>accelerated-load</tt>
option. For a complete list of the command-line options run:

    gospeccy -help

If you can't wait to see what this machine can do, try the nice
[Fire104b](http://pouet.net/prod.php?which=54076) intro by Andrew
Gerrand included in the gospeccy distribution! In the gospeccy folder,
run:

    gospeccy -2x snapshots/Syntax09nF.z80

To automatically download a program from [World of spectrum](http://www.worldofspectrum.org),
and start it:

    gospeccy -wos="horace*tower"

For more, try searching the Internet for ZX Spectrum 48k
games and demos in Z80 and TAP format. For example:

* [World of spectrum](http://www.worldofspectrum.org/archive.html)
* [Pouet.net search](http://pouet.net/prodlist.php?platform[]=ZX+Spectrum)
* [Forever](http://forever.zeroteam.sk/download.htm)

# Key bindings

    Host computer   ZX Spectrum
    ---------------------------
    CTRL            SYMBOL SHIFT
    LEFT SHIFT      CAPS SHIFT
    [a-z0-9]        [A-Z0-9]
    SPACE           SPACE

For more info about key bindings see file <tt>spectrum/keyboard.go</tt>

# Proprietary games and system ROM

Generally, games/programs are protected by copyright so none of them
is included in GoSpeccy. However, it is possible to find tons of games for
the ZX Spectrum on the Internet. The system ROM for Spectrum 48k can be freely
distributed and so it's included in the GoSpeccy distribution.

# Convention over Configuration

Loading files in the emulator relies on a Convention over
Configuration approach. To enjoy it, you should create the following
folder structure:

<pre>
mkdir -p $HOME/.config/gospeccy/roms	# System roms folder
mkdir -p $HOME/.config/gospeccy/scripts	# Scripts folder
</pre>

In the scripts folder, create file `config_local.go` with the following
contents:

<pre>
// Search path for snapshots, scripts, etc
addSearchPath("/home/user/gospeccy")
</pre>

After this, to load `/home/user/gospeccy/somegame.z80` simply execute:

<pre>
gospeccy somegame.z80
</pre>

# Screenshots

Manic Miner running on GoSpeccy.

![Manic Miner running on GoSpeccy](http://sites.google.com/site/remogatto/gospeccy_running_scaled.png)

# To Do

* Fix some memory and I/O contention bugs
* Add support for more file formats (take a look [here](http://www.worldofspectrum.org/faq/reference/formats.htm))
* Add support for tape saving
* Better general performance
* Add more filters and improve the scaler
* Add new graphical backends (Go's exp/draw?)

# Credits

* Thanks to [⚛](http://github.com/0xe2-0x9a-0x9b) for giving a new
  whole direction to this project.
* Thanks to the people on
  [golang-nuts](http://groups.google.com/group/golang-nuts) for giving
  feedback and support.
* Thanks to Andrew Gerrand for the crackling Fire104b demo.

# Contacts

* andrea.fazzi@alcacoop.it
* http://twitter.com/remogatto
* http://www.facebook.com/remochat
* http://remogatto.github.com/

# License

Copyright (c) 2010 Andrea Fazzi

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



ChangeLog
0.8.0 / 2012-02-22
==================

	* Use the 'go' tool to build and install GoSpeccy

	* Enable WorldOfSpectrum searches and downloads ("-wos" command-line
	  option, wos* script functions). Info about a WorldOfSpectrum file
	  includes publication type, machine type, and rating.

	* New script functions: vars(), wosFind(string), wosDownload(WOS),
	  wosLoad(WOS)

	* Evaluating an expression in the console prints the result of the
	  expression

	* Enable compilation without SDL (this is an experimental feature)

0.7.0 / 2011-05-21
==================
	* Use GOAM [http://github.com/0xe2-0x9a-0x9b/goam] to build and
	  install GoSpeccy

	* Use Clingon [http://github.com/remogatto/clingon] as the
	  command-line GUI, instead of a C readline wrapper. The readline
	  support is gone.

	* Use [http://github.com/remogatto/prettytest], instead of an internal
	  copy of the PrettyTest package

	* Improved audio resampling method which usually produces less noise

	* Added command-line options: "-hostcpu-profile", "-show-paint",
	  "-audio", "-audio-freq", "-audio-hq"

	* New script functions: cmdLineArg() string, fullscreen(bool),
	  showPaint(bool), audioFreq(uint), audioHQ(bool)

	* Renamed script functions: sound(bool) --> audio(bool)

	* Performance optimizations: store events events in an array (instead
	  of storing them in a linked-list), do not try to use uninitialized
	  Go  variables, lower number of conversions between numeric types
	  with  different bit-lengths, do not allocate memory when computing
	  the address of a pixel in a SDL surface, do not allocate memory when
	  executing certain instructions (INC, DEC, POP, RET, and others),
	  use Go arrays instead of 'container/vector/Vector', read the ROM
	  file only once instead of reading it on each reset, faster typing
	  of [LOAD ""] when loading from tape

	* Better display refresh stability thanks to reduced memory
	  allocation. That is: less reasons for the Go runtime to stop the
	  program to perform memory garbage-collection.

	* Make the Z80 CPU conformance test complete much faster

	* Add initial support for Kempston joystick

	* Automatically detect when to pause/unpause the tape.
	  This enables loading of multi-part TAP files, and of programs
	  that expect the audio data to contain periods of silence.

	* Play 8063 pulses on all TAP headers

	* Limit the maximum speedup of accelerated loading to 20

	* Enable loading of the "VeraMono.ttf" font from "$HOME/.gospeccy"
	  or from the installation directory

	* Do not initialize SDL if GoSpeccy is unable to detect the type of
	  the program on the command-line

	* Changing the FPS should now result in fever sound buffer underflows

	* Do not set GOMAXPROCS if there is an environment variable named
	  'GOMAXPROCS'

	* Disable writes to ZX Spectrum ROM (to the first 0x4000 bytes of
	  memory). As a side-effect it enables "RANDOMIZE USR 0" to work
	  correctly.

	* Change the script help text so that it fits the width of the console
	  when the scale is set to 2

	* Add Go-SDL TTF version check. Print video surface resolution if in
	  verbose mode

	* Update to a newer version of Go-PerfEvents that allows GoSpeccy
	  to potentially compile&run in a wider range of software environments

	* Fix a couple of real&potential bugs, deadlocks and race conditions

	* Minor updates to the README.md file

0.6.1 / 2010-11-12
==================
	* Add support for accelerated tape loading both in console and command-line

0.6.0 / 2010-11-11
==================
	* Add support for loading tapes in TAP format through the system
	ROM loading routine

	* Add support for Z80 48k snapshots (all Z80 versions)

	* Enable loading of a program (tape or snapshot) from a ZIP
	archive

	* Add system/blackbox tests (see src/test)

	* Add puts() function to the console Test script() console
	function

	* Create a separate package for the console

	* Faster emulation of memory contention

	* Enable a script or the console to add a new search path (handful
	if used in the "startup_local.go" script), a snapshot specified as
	an argument to GoSpeccy now superseeds any snapshots loaded from
	the startup script

	* Inline code that is used to set individual pixels (for up to 2x
	rendering performance improvement)

	* Improve conversion of 2-bit beeper levels to 16-bit audio (fixes
	the music loudness in Manic Miner)

	* Fix emulation of HALT instruction

	* Fix many memory and I/O contention issues (checked against
	JSpeccy 0.87, and against a couple of timing tests)

	* Accurate emulation of the border (Overscan Demo by Busy Soft
	works perfectly)

	* Border height is now 32 pixels instead of 24

	* Improve simulation of RETN when loading a SNA snapshot

	* Reduce visual artifacting caused by non-atomic SDL surface
	updates

	* Fixed border rendering when a frame is not sent to the rendering
	backend

0.5.1 / 2010-09-18
==================

        * Add an audio stress-test, improve the application shutdown procedure

0.5.0 / 2010-09-18
==================

	* Hide mouse cursor in fullscreen mode

	* Enable uninstallation

	* Allow testing and benchmarking in the "src/spectrum" directory
	without the need to install the C readline wrapper

	* Add convention over configuration for loading snapshots/roms/scripts

	* Add convention over configuration when loading snapshots and
	system rom(s). See #16. Modify the main Makefile in order to copy
	ths system rom(s) in $(GOROOT)/pkg/$(GOOS)_$(GOARCH)/gospeccy/roms
	directory

	* Add support for loading and running Go scripts, add a simple
	display stress-test script.

	* Check Go-SDL version descriptors to prevent running with an
	incompatible Go-SDL version

	* Console & readline: handle terminal resize (SIGWINCH signal)

	* If video/audio cannot be initialized, exit with an error message (instead of calling panic())

	* Add beeper sound support

	* Use ⚛SDL's event channel.

	* Prevent the display backend from blocking CPU emulation.


src/console.go
package main

import (
	"spectrum"
	"spectrum/formats"
	"clingon"
	"exp/eval"
	"fmt"
	"io/ioutil"
	"os"
	"container/vector"
	"bytes"
	"strings"
	"time"
)

// ==============
// Some variables
// ==============

// These variables are set only once, before starting new goroutines,
// so there is no need for controlling concurrent access via a sync.Mutex
var app *spectrum.Application
var speccy *spectrum.Spectrum48k
var i *Interpreter
var w *eval.World

var IgnoreStartupScript = false

const SCRIPT_DIRECTORY = "scripts"
const STARTUP_SCRIPT = "startup"

type Interpreter struct {
	console *clingon.Console
}

func (i *Interpreter) SetConsole(console *clingon.Console) {
	i.console = console
}

func (i *Interpreter) Run(command string) os.Error {
	return i.run(w, command)
}

func (i *Interpreter) Print(str string) {
	i.console.PushLines(strings.Split(str, "\n", -1))
}

// ================
// Various commands
// ================

var help_keys vector.StringVector
var help_vals vector.StringVector

func (i *Interpreter) printHelp() {
	var buf bytes.Buffer

	fmt.Fprintf(&buf, "\nAvailable commands:\n")

	maxKeyLen := 1
	for i := 0; i < help_keys.Len(); i++ {
		if len(help_keys[i]) > maxKeyLen {
			maxKeyLen = len(help_keys[i])
		}
	}

	for i := 0; i < help_keys.Len(); i++ {
		fmt.Fprintf(&buf, "    %s", help_keys[i])
		for j := len(help_keys[i]); j < maxKeyLen; j++ {
			fmt.Fprintf(&buf, " ")
		}
		fmt.Fprintf(&buf, "  %s\n", help_vals[i])
	}

	i.Print(buf.String())
}

// Signature: func help()
func wrapper_help(t *eval.Thread, in []eval.Value, out []eval.Value) {
	i.printHelp()
}

// Signature: func exit()
func wrapper_exit(t *eval.Thread, in []eval.Value, out []eval.Value) {
	// Implementation note:
	//   The following test has to be there only in cases in which something can go wrong.
	//   For example if the user would try to execute "exit(); sound(false)" then GoSpeccy would panic.
	//   An alternative way would be to actually terminate the whole program at the 1st statement - so that
	//   "sound(false)" or whatever is not executed - alas this is somewhat problematic,
	//   since once the script "exit(); sound(false)" runs, it cannot be stopped halfway
	//   through its execution. Using "runtime.Goexit()" would solve this issue, but only partially,
	//   since it is potentially possible for the statement "sound(false)" to be hidden in a defer statement.
	//   So, the best option (until somebody implements a better one) is to convert the problematic commands
	//   into statements that are doing nothing while the application is in the process of being exited.
	if app.TerminationInProgress() {
		return
	}

	app.RequestExit()
}

// Signature: func reset()
func wrapper_reset(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	speccy.CommandChannel <- spectrum.Cmd_Reset{make(chan bool, 1)}
}

// Signature: func addSearchPath(path string)
func wrapper_addSearchPath(t *eval.Thread, in []eval.Value, out []eval.Value) {
	path := in[0].(eval.StringValue).Get(t)
	spectrum.AddCustomSearchPath(path)
}

// Signature: func load(path string)
func wrapper_load(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	path := in[0].(eval.StringValue).Get(t)

	path = spectrum.ProgramPath(path)

	var program interface{}
	program, err := formats.ReadProgram(path)
	if err != nil {
		i.Print(err.String())
		return
	}

	if _, isTAP := program.(*formats.TAP); isTAP {
		romLoaded := make(chan bool, 1)
		speccy.CommandChannel <- spectrum.Cmd_Reset{romLoaded}
		<-romLoaded
	}

	errChan := make(chan os.Error)
	speccy.CommandChannel <- spectrum.Cmd_Load{path, program, errChan}

	err = <-errChan
	if err != nil {
		i.Print(err.String())
	}
}

// Signature: func save(path string)
func wrapper_save(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	path := in[0].(eval.StringValue).Get(t)

	ch := make(chan *formats.FullSnapshot)
	speccy.CommandChannel <- spectrum.Cmd_MakeSnapshot{ch}

	fullSnapshot := <-ch

	data, err := fullSnapshot.EncodeSNA()
	if err != nil {
		i.Print(err.String())
		return
	}

	err = ioutil.WriteFile(path, data, 0600)
	if err != nil {
		i.Print(err.String())
	}

	if app.Verbose {
		i.Print(fmt.Sprintf("wrote SNA snapshot \"%s\"", path))
	}
}

// Signature: func scale(n uint)
func wrapper_scale(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	n := in[0].(eval.UintValue).Get(t)

	switch n {
	case 1:
		finished := make(chan byte)
		speccy.CommandChannel <- spectrum.Cmd_CloseAllDisplays{finished}
		<-finished

		speccy.CommandChannel <- spectrum.Cmd_AddDisplay{spectrum.NewSDLScreen(app)}

	case 2:
		finished := make(chan byte)
		speccy.CommandChannel <- spectrum.Cmd_CloseAllDisplays{finished}
		<-finished

		speccy.CommandChannel <- spectrum.Cmd_AddDisplay{spectrum.NewSDLScreen2x(app, /*fullscreen*/ false)}
	}
}

// Signature: func fps(n float)
func wrapper_fps(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	fps := in[0].(eval.FloatValue).Get(t)
	speccy.CommandChannel <- spectrum.Cmd_SetFPS{float(fps)}
}

// Signature: func ula_accuracy(accurateEmulation bool)
func wrapper_ulaAccuracy(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	accurateEmulation := in[0].(eval.BoolValue).Get(t)
	speccy.CommandChannel <- spectrum.Cmd_SetUlaEmulationAccuracy{accurateEmulation}
}

// Signature: func sound(enable bool)
func wrapper_sound(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	enable := in[0].(eval.BoolValue).Get(t)

	if enable {
		audio, err := spectrum.NewSDLAudio(app)
		if err == nil {
			finished := make(chan byte)
			speccy.CommandChannel <- spectrum.Cmd_CloseAllAudioReceivers{finished}
			<-finished

			speccy.CommandChannel <- spectrum.Cmd_AddAudioReceiver{audio}
		} else {
			i.Print(err.String())
		}
	} else {
		finished := make(chan byte)
		speccy.CommandChannel <- spectrum.Cmd_CloseAllAudioReceivers{finished}
		<-finished
	}
}

// Signature: func wait(milliseconds uint)
func wrapper_wait(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	milliseconds := uint(in[0].(eval.UintValue).Get(t))
	time.Sleep(1e6 * int64(milliseconds))
}

// Signature: func script(scriptName string)
func wrapper_script(t *eval.Thread, in []eval.Value, out []eval.Value) {
	scriptName := in[0].(eval.StringValue).Get(t)

	err := i.runScript(w, scriptName, /*optional*/ false)
	if err != nil {
		i.Print(err.String())
		return
	}
}

// Signature: func optionalScript(scriptName string)
func wrapper_optionalScript(t *eval.Thread, in []eval.Value, out []eval.Value) {
	scriptName := in[0].(eval.StringValue).Get(t)

	err := i.runScript(w, scriptName, /*optional*/ true)
	if err != nil {
		i.Print(err.String())
		return
	}
}

// Signature: func screenshot(screenshotName string)
func wrapper_screenshot(t *eval.Thread, in []eval.Value, out []eval.Value) {
	if app.TerminationInProgress() {
		return
	}

	path := in[0].(eval.StringValue).Get(t)

	ch := make(chan []byte)
	speccy.CommandChannel <- spectrum.Cmd_MakeVideoMemoryDump{ch}

	data := <-ch

	err := ioutil.WriteFile(path, data, 0600)

	if err != nil {
		i.Print(err.String())
	}

	if app.Verbose {
		i.Print(fmt.Sprintf("wrote screenshot \"%s\"", path))
	}
}

// Signature: func puts(str string)
func wrapper_puts(t *eval.Thread, in []eval.Value, out []eval.Value) {
	str := in[0].(eval.StringValue).Get(t)
	i.Print(str)
}

// Signature: func acceleratedLoad(on bool)
func wrapper_acceleratedLoad(t *eval.Thread, in []eval.Value, out []eval.Value) {
	value := in[0].(eval.BoolValue).Get(t)
	speccy.EnableAcceleratedLoad(value)
}

// ==============
// Initialization
// ==============

func defineFunctions(w *eval.World) {
	{
		var functionSignature func()
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_help, functionSignature)
		w.DefineVar("help", funcType, funcValue)
		help_keys.Push("help()")
		help_vals.Push("This help")
	}

	{
		var functionSignature func()
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_exit, functionSignature)
		w.DefineVar("exit", funcType, funcValue)
		help_keys.Push("exit()")
		help_vals.Push("Terminate this program")
	}

	{
		var functionSignature func()
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_reset, functionSignature)
		w.DefineVar("reset", funcType, funcValue)
		help_keys.Push("reset()")
		help_vals.Push("Reset the emulated machine")
	}

	{
		var functionSignature func(string)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_addSearchPath, functionSignature)
		w.DefineVar("addSearchPath", funcType, funcValue)
		help_keys.Push("addSearchPath(path string)")
		help_vals.Push("Append a path to the list of paths searched when loading snapshots")
	}

	{
		var functionSignature func(string)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_load, functionSignature)
		w.DefineVar("load", funcType, funcValue)
		help_keys.Push("load(path string)")
		help_vals.Push("Load state from file (.SNA, .Z80, .Z80.ZIP, etc)")
	}

	{
		var functionSignature func(string)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_save, functionSignature)
		w.DefineVar("save", funcType, funcValue)
		help_keys.Push("save(path string)")
		help_vals.Push("Save state to file (SNA format)")
	}

	{
		var functionSignature func(uint)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_scale, functionSignature)
		w.DefineVar("scale", funcType, funcValue)
		help_keys.Push("scale(n uint)")
		help_vals.Push("Change the display scale")
	}

	{
		var functionSignature func(float)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_fps, functionSignature)
		w.DefineVar("fps", funcType, funcValue)
		help_keys.Push("fps(n float)")
		help_vals.Push("Change the display refresh frequency")
	}

	{
		var functionSignature func(bool)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_ulaAccuracy, functionSignature)
		w.DefineVar("ula_accuracy", funcType, funcValue)
		help_keys.Push("ula_accuracy(accurateEmulation bool)")
		help_vals.Push("Enable/disable accurate emulation of screen bitmap and screen attributes")
	}

	{
		var functionSignature func(bool)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_sound, functionSignature)
		w.DefineVar("sound", funcType, funcValue)
		help_keys.Push("sound(enable bool)")
		help_vals.Push("Enable or disable sound")
	}

	{
		var functionSignature func(uint)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_wait, functionSignature)
		w.DefineVar("wait", funcType, funcValue)
		help_keys.Push("wait(milliseconds uint)")
		help_vals.Push("Wait the specified amount of time before issuing the next command")
	}

	{
		var functionSignature func(string)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_script, functionSignature)
		w.DefineVar("script", funcType, funcValue)
		help_keys.Push("script(scriptName string)")
		help_vals.Push("Load and evaluate the specified Go script")
	}

	{
		var functionSignature func(string)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_optionalScript, functionSignature)
		w.DefineVar("optionalScript", funcType, funcValue)
		help_keys.Push("optionalScript(scriptName string)")
		help_vals.Push("Load (if found) and evaluate the specified Go script")
	}

	{
		var functionSignature func(string)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_screenshot, functionSignature)
		w.DefineVar("screenshot", funcType, funcValue)
		help_keys.Push("screenshot(screenshotName string)")
		help_vals.Push("Take a screenshot of the current display")
	}
	{
		var functionSignature func(string)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_puts, functionSignature)
		w.DefineVar("puts", funcType, funcValue)
		help_keys.Push("puts(str string)")
		help_vals.Push("Print the given string")
	}

	{
		var functionSignature func(bool)
		funcType, funcValue := eval.FuncFromNativeTyped(wrapper_acceleratedLoad, functionSignature)
		w.DefineVar("acceleratedLoad", funcType, funcValue)
		help_keys.Push("acceleratedLoad(on bool)")
		help_vals.Push("Set accelerated tape load on/off")
	}

}

// Runs the specified Go source code in the context of 'w'
func (i *Interpreter) run(w *eval.World, sourceCode string) os.Error {
	var err os.Error
	var code eval.Code

	code, err = w.Compile(sourceCode)
	if err != nil {
		i.Print(err.String())
		return err
	}

	_, err = code.Run()
	if err != nil {
		i.Print(err.String())
		return err
	}

	return err
}

// Loads and evaluates the specified Go script
func (i *Interpreter) runScript(w *eval.World, scriptName string, optional bool) os.Error {
	fileName := scriptName + ".go"
	data, err := ioutil.ReadFile(spectrum.ScriptPath(fileName))
	if err != nil {
		if !optional {
			return err
		} else {
			return nil
		}
	}

	var buf bytes.Buffer
	buf.Write(data)
	i.run(w, buf.String())

	return nil
}

func InitConsole(_app *spectrum.Application, _speccy *spectrum.Spectrum48k) {
	if app != nil {
		panic("running multiple consoles is unsupported")
	}

	app = _app
	speccy = _speccy
	w = eval.NewWorld()

	defineFunctions(w)

	// Run the startup script
	var err os.Error

	err = i.runScript(w, STARTUP_SCRIPT, /*optional*/ IgnoreStartupScript)
	if err != nil {
		app.PrintfMsg("%s", err)
		app.RequestExit()
		return
	}
}

// Lines below will be uncommented when/if the keypress console
// command will be implemented.

// type uintV uint

// func newUint(v uint) *uintV {
// 	vp := uintV(v)
// 	return &vp
// }


// func (v *uintV) String() string { return fmt.Sprint(*v) }
// func (v *uintV) Assign(t *eval.Thread, o eval.Value) { *v = uintV(o.(eval.UintValue).Get(t)) }
// func (v *uintV) Get(*eval.Thread) uint64 { return uint64(*v) }
// func (v *uintV) Set(t *eval.Thread, x uint64) { *v = uintV(x) }

// func defineConstants(w *eval.World) {
// 	w.DefineConst("KEY_1", eval.UintType, newUint(0))
// 	w.DefineConst("KEY_2", eval.UintType, newUint(1))
// 	w.DefineConst("KEY_3", eval.UintType, newUint(2))
// 	w.DefineConst("KEY_4", eval.UintType, newUint(3))
// 	w.DefineConst("KEY_5", eval.UintType, newUint(4))
// 	w.DefineConst("KEY_6", eval.UintType, newUint(5))
// 	w.DefineConst("KEY_7", eval.UintType, newUint(6))
// 	w.DefineConst("KEY_8", eval.UintType, newUint(7))
// 	w.DefineConst("KEY_9", eval.UintType, newUint(8))
// 	w.DefineConst("KEY_0", eval.UintType, newUint(9))
// 	w.DefineConst("KEY_Q", eval.UintType, newUint(10))
// 	w.DefineConst("KEY_W", eval.UintType, newUint(11))
// 	w.DefineConst("KEY_E", eval.UintType, newUint(12))
// 	w.DefineConst("KEY_R", eval.UintType, newUint(13))
// 	w.DefineConst("KEY_T", eval.UintType, newUint(14))
// 	w.DefineConst("KEY_Y", eval.UintType, newUint(15))
// 	w.DefineConst("KEY_U", eval.UintType, newUint(16))
// 	w.DefineConst("KEY_I", eval.UintType, newUint(17))
// 	w.DefineConst("KEY_O", eval.UintType, newUint(18))
// 	w.DefineConst("KEY_P", eval.UintType, newUint(19))

// 	w.DefineConst("KEY_A", eval.UintType, newUint(20))
// 	w.DefineConst("KEY_S", eval.UintType, newUint(21))
// 	w.DefineConst("KEY_D", eval.UintType, newUint(22))
// 	w.DefineConst("KEY_F", eval.UintType, newUint(23))
// 	w.DefineConst("KEY_G", eval.UintType, newUint(24))
// 	w.DefineConst("KEY_H", eval.UintType, newUint(25))
// 	w.DefineConst("KEY_J", eval.UintType, newUint(26))
// 	w.DefineConst("KEY_K", eval.UintType, newUint(27))
// 	w.DefineConst("KEY_L", eval.UintType, newUint(28))
// 	w.DefineConst("KEY_Enter", eval.UintType, newUint(29))

// 	w.DefineConst("KEY_CapsShift", eval.UintType, newUint(30))
// 	w.DefineConst("KEY_Z", eval.UintType, newUint(31))
// 	w.DefineConst("KEY_X", eval.UintType, newUint(32))
// 	w.DefineConst("KEY_C", eval.UintType, newUint(33))
// 	w.DefineConst("KEY_V", eval.UintType, newUint(34))
// 	w.DefineConst("KEY_B", eval.UintType, newUint(35))
// 	w.DefineConst("KEY_N", eval.UintType, newUint(36))
// 	w.DefineConst("KEY_M", eval.UintType, newUint(37))
// 	w.DefineConst("KEY_SymbolShift", eval.UintType, newUint(38))
// 	w.DefineConst("KEY_Space", eval.UintType, newUint(39))
// }
